lib/miq_pglogical.rb

```diff
@@ -7,21 +7,34 @@ class MiqPglogical
   PUBLICATION_NAME = 'miq'.freeze
   ALWAYS_EXCLUDED_TABLES = %w(ar_internal_metadata schema_migrations repl_events repl_monitor repl_nodes).freeze
 
-  def self.with_connection_error_handling
-    retry_attempted ||= false
-    yield
-  rescue PG::ConnectionBad
-    raise if retry_attempted
-
-    pglogical(true)
-    retry_attempted = true
-    retry
-  end
+  module ConnectionHandling
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      def with_connection_error_handling
+        retry_attempted ||= false
+        yield
+      rescue PG::ConnectionBad
+        raise if retry_attempted
+
+        pglogical(true)
+        retry_attempted = true
+        retry
+      end
+
+      def pglogical(refresh = false)
+        @pglogical = nil if refresh
+        @pglogical ||= PG::LogicalReplication::Client.new(pg_connection)
```

Comment:
This is where `pg_connection` is implemented, and makes sure it calls the one that is explicitly defined in each class.